### ISSUE TITLE
TypedDict keys reuse?

### ISSUE THREAD

**Issue by jstasiak** [python/mypy#4128]
I may be in a need of the following - two ``TypeDict``s with the same keys, one of them created with ``total=False``:

```python
from mypy_extensions import TypedDict

Details = TypedDict('Details', {'name': str, 'age': int, address: str})
DetailsSubset = TypedDict('DetailsSubset', {'name': str, 'age': int, address: str}, total=False)
```

Is there a way to reuse the keys instead of having to repeat the ``{'name': str, 'age': int, address: str}`` fragment twice?

**Comment by ilevkivskyi** [comment]
This is currently not possible. We will probably need the ``Const`` type modifier for this to work, see https://github.com/python/mypy/issues/1214

**Comment by JukkaL** [comment]
We now have a `Final` modified but that isn't enough, since we never implicitly infer TypedDict types from dict expressions.

**Comment by ilevkivskyi** [comment]
Another4 request for this appeared, see https://github.com/python/mypy/issues/8186, so raising priority to normal.

**Comment by ilevkivskyi** [comment]
(Just to clarify, we can start from just support final names for individual keys, which should be pretty easy.)

**Comment by cwhy** [comment]
@ilevkivskyi  I agree that we can only support final names for individual keys, as stated at the end of PEP-591:

> Type checkers should treat uses of a final name that was initialized with a literal as if it was replaced by the literal. For example, the following should be allowed:

> ```from typing import NamedTuple, Final
> 
> X: Final = "x"
> Y: Final = "y"
> N = NamedTuple("N", [(X, int), (Y, int)])
> ```

It is sometimes useful to have different types on the same label as well:

```
X: Final = "x"
Y: Final = "y"
Position = NamedTuple("Position", [(X, int), (Y, int)])
DimensionName = NamedTuple("DimensionName", [(X, str), (Y, str)])
```

**Comment by tony** [comment]
_Let me know if you'd like me to remove this and create a separate issue if it's off scope_

Allow all "typed fields" / keyed data structures to reuse their typings.

e.g. [`typing.TypedDict`](https://docs.python.org/3/library/typing.html#typing.TypedDict), [`typing.NamedTuple`](https://docs.python.org/3/library/typing.html#typing.NamedTuple) and [`dataclasses.dataclass`](https://docs.python.org/3/library/dataclasses.html#dataclasses.dataclass).

## Scenario: Typed exports from dataclass

Assume my case, I want to offer users ability to export a typed tuple and dict from my `Details` dataclass, [`dataclasses.astuple`](https://docs.python.org/3/library/dataclasses.html#dataclasses.astuple) and [`dataclasses.asdict`](https://docs.python.org/3/library/dataclasses.html#dataclasses.asdict)

```python
#!/usr/bin/env python
import dataclasses
from typing import NamedTuple, TypedDict, get_type_hints


class DetailsDict(TypedDict):
    name: str
    age: int
    address: str


class DetailsTuple(NamedTuple):
    name: str
    age: int
    address: str


@dataclasses.dataclass
class Details:
    name: str
    age: int
    address: str

    def to_dict(self) -> DetailsDict:
        # return dataclasses.asdict(self, dict_factory=DetailsDict)
        return DetailsDict(**dataclasses.asdict(self))

    def to_tuple(self) -> DetailsTuple:
        # return dataclasses.astuple(self, tuple_factory=DetailsTuple)
        return DetailsTuple(*dataclasses.astuple(self))


john = Details(name="John", age=25, address="123 Address St")
print(john)
print(john.to_dict())
print(john.to_tuple())
print(get_type_hints(john))
print(get_type_hints(john.to_dict))
print(get_type_hints(john.to_tuple))
```

Output:

```
Details(name='John', age=25, address='123 Address St')
{'name': 'John', 'age': 25, 'address': '123 Address St'}
DetailsTuple(name='John', age=25, address='123 Address St')
{'name': <class 'str'>, 'age': <class 'int'>, 'address': <class 'str'>}
{'return': <class '__main__.DetailsDict'>}
{'return': <class '__main__.DetailsTuple'>
```

We need a way to reuse the field typings across the different keyed data structures.
